src/Mutation/MutationFactory.php

```diff
@@ -0,0 +1,220 @@
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+declare(strict_types=1);
+
+namespace Infection\Mutation;
+
+use function array_intersect_key;
+use function implode;
+use Infection\AbstractTestFramework\Coverage\TestLocation;
+use Infection\Differ\Differ;
+use Infection\PhpParser\MutatedNode;
+use function md5;
+use PhpParser\Node;
+use PhpParser\PrettyPrinterAbstract;
+use function Safe\array_flip;
+use function Safe\sprintf;
+use Webmozart\Assert\Assert;
+
+/**
+ * @internal
+ * @final
+ */
+class MutationFactory
+{
+    private $tmpDir;
+    private $differ;
+    private $printer;
+
+    /**
+     * @var string[]
+     */
+    private $printedFileCache = [];
+    private $mutationCodeFactory;
+
+    public function __construct(
+        string $tmpDir,
+        Differ $differ,
+        PrettyPrinterAbstract $printer,
+        MutationCodeFactory $mutationCodeFactory
+    ) {
+        $this->tmpDir = $tmpDir;
+        $this->differ = $differ;
+        $this->printer = $printer;
+        $this->mutationCodeFactory = $mutationCodeFactory;
+    }
+
+    /**
+     * @param Node[] $originalFileAst
+     * @param array<string|int|float> $attributes
+     * @param class-string $mutatedNodeClass
+     * @param TestLocation[] $tests
+     */
+    public function create(
+        string $originalFilePath,
+        array $originalFileAst,
+        string $mutatorName,
+        array $attributes,
+        string $mutatedNodeClass,
+        MutatedNode $mutatedNode,
+        int $mutationByMutatorIndex,
+        array $tests
+    ): Mutation {
+        foreach (MutationAttributeKeys::ALL as $key) {
+            Assert::keyExists($attributes, $key);
+        }
+
+        $attributes = array_intersect_key($attributes, array_flip(MutationAttributeKeys::ALL));
+
+        return new Mutation(
+            $originalFilePath,
+            $mutatorName,
+            (int) $attributes[MutationAttributeKeys::START_LINE],
+            $tests,
+            function () use (
+                $originalFilePath,
+                $originalFileAst,
+                $mutatorName,
+                $attributes,
+                $mutatedNodeClass,
+                $mutatedNode,
+                $mutationByMutatorIndex
+            ): MutationCalculatedState {
+                return $this->calculateState(
+                    $originalFilePath,
+                    $originalFileAst,
+                    $mutatorName,
+                    $attributes,
+                    $mutatedNodeClass,
+                    $mutatedNode,
+                    $mutationByMutatorIndex
+                );
```

Comment:
Why a callback here is better than a DI we use elsewhere?

Reply:
for laziness. An alternative would be to either:

- calculate everything upfront, but then we would have a performance regression
- inject the required services in `Mutation` and push back the complexity of calculating the fields in `Mutation`

The later is the most sensible option if you don't like this closure, but I found the closure way easier to handle (and in the tests as well since you can just create a dummy calculated state, you don't require any services)

Reply:
How much gain do you think we get from this additional laziness?

Reply:
Can give it a go but I suspect 5-10%.

Note that it won't change much IMO:

- in the factory you pass the values directly instead of the closure, ok this part is simpler
- in `Mutation` you pass the values directly instead of the closure, ok a bit simpler
- everywhere else: not much difference

The closure returns a simple object, so instead of having:

```php
static function () {
     return new MutationCalculatedState('hash', 'changedCode', 'diff');
}
```

You will have:

```php
'hash', 
'changedCode', 
'diff'
```

Reply:
+12% https://blackfire.io/profiles/compare/37e0648d-2fd7-4b21-92ba-13797432a618/graph

Reply:
> but I suspect 5-10%.

I did some benchmarks, and didn't notice anything like that. More like to the contrary. If you can find a proof there's a positive impact, it'd be great. Else we'd complicating things without a reason.

Reply:
- the hash is used only if you create a process, which we skip for uncovered mutations
- the diff for now is used only for the reporting, so it depends on the reports used
- same for mutated code

Reply:
In other words: if you use the debug report where we display the diff regardless of the status of the mutation, then it won't make a difference indeed.

In any other scenario, it will

Reply:
Removed for now